File: security-platform/agents/node/src/agent.ts
import { NodeSDK } from '@opentelemetry/sdk-trace-node';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { OTLPTraceExporter } from '@opentelemetry/exporter-otlp-http';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { AgentConfig } from './types';
import { Redactor } from './redaction';
import { PolicyEngine } from './policy';

export class Agent {
  private sdk: NodeSDK | null = null;
  private config: AgentConfig;
  private redactor: Redactor;
  private policyEngine: PolicyEngine;

  constructor(config: AgentConfig) {
    this.config = config;
    this.redactor = new Redactor(config.redactionRules || []);
    this.policyEngine = new PolicyEngine(config.policy || { mode: 'observe' });
  }

  start(): void {
    const resource = new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: this.config.serviceName || 'node-service',
      [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: this.config.environment || 'production',
    });

    const exporter = new OTLPTraceExporter({
      url: this.config.otlpEndpoint || `${this.config.controlPlaneUrl}/v1/traces`,
      headers: this.config.authKey ? { Authorization: `Bearer ${this.config.authKey}` } : {},
    });

    this.sdk = new NodeSDK({
      resource,
      traceExporter: exporter,
      instrumentations: [getNodeAutoInstrumentations()],
      spanProcessor: new BatchSpanProcessor(exporter, {
        maxQueueSize: this.config.telemetry?.maxQueueSize || 2048,
        maxExportBatchSize: this.config.telemetry?.batchSize || 100,
        scheduledDelayMillis: this.parseDuration(this.config.telemetry?.batchTimeout || '5s'),
        exportTimeoutMillis: this.parseDuration(this.config.telemetry?.exportTimeout || '30s'),
      }),
    });

    this.sdk.start();
  }

  stop(): void {
    this.sdk?.shutdown();
  }

  redact(data: any): any {
    return this.redactor.redact(data);
  }

  checkPolicy(action: string, context: Record<string, any>): { allowed: boolean; reason?: string } {
    return this.policyEngine.check(action, context);
  }

  private parseDuration(duration: string): number {
    const match = duration.match(/^(\d+)([smh])$/);
    if (!match) return 5000; // default 5s

    const value = parseInt(match[1], 10);
    const unit = match[2];

    switch (unit) {
      case 's':
        return value * 1000;
      case 'm':
        return value * 60 * 1000;
      case 'h':
        return value * 60 * 60 * 1000;
      default:
        return 5000;
    }
  }
}

File: security-platform/agents/go/agent.go
package agent

import (
	"context"
	"fmt"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	"go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"go.opentelemetry.io/otel/trace"
)

// Agent represents the Security Platform Agent.
type Agent struct {
	config       *Config
	tracerProvider *trace.TracerProvider
	redactor     *Redactor
	policyEngine *PolicyEngine
	started      bool
}

// Config holds the agent configuration.
type Config struct {
	ControlPlaneURL string
	AuthKey         string
	ServiceName     string
	Environment     string
	OtlpEndpoint    string
	Telemetry       TelemetryConfig
	RedactionRules  []string
	Policy          PolicyConfig
}

// TelemetryConfig holds telemetry configuration.
type TelemetryConfig struct {
	BatchSize      int
	BatchTimeout   time.Duration
	ExportTimeout  time.Duration
	MaxQueueSize   int
}

// PolicyConfig holds policy configuration.
type PolicyConfig struct {
	Mode string // "observe" or "block"
}

// NewAgent creates a new agent instance.
func NewAgent(config *Config) *Agent {
	if config == nil {
		config = &Config{}
	}

	// Set defaults from environment variables
	if config.ControlPlaneURL == "" {
		config.ControlPlaneURL = os.Getenv("SECURITY_PLATFORM_CONTROL_PLANE_URL")
		if config.ControlPlaneURL == "" {
			config.ControlPlaneURL = "https://api.securityplatform.com"
		}
	}

	if config.AuthKey == "" {
		config.AuthKey = os.Getenv("SECURITY_PLATFORM_AUTH_KEY")
	}

	if config.ServiceName == "" {
		config.ServiceName = "go-service"
	}

	if config.Environment == "" {
		config.Environment = "production"
	}

	if config.Telemetry.BatchSize == 0 {
		config.Telemetry.BatchSize = 100
	}

	if config.Telemetry.BatchTimeout == 0 {
		config.Telemetry.BatchTimeout = 5 * time.Second
	}

	if config.Telemetry.ExportTimeout == 0 {
		config.Telemetry.ExportTimeout = 30 * time.Second
	}

	if config.Telemetry.MaxQueueSize == 0 {
		config.Telemetry.MaxQueueSize = 2048
	}

	if config.Policy.Mode == "" {
		config.Policy.Mode = "observe"
	}

	return &Agent{
		config:       config,
		redactor:     NewRedactor(config.RedactionRules),
		policyEngine: NewPolicyEngine(&config.Policy),
	}
}

// Start initializes and starts the agent.
func (a *Agent) Start() error {
	if a.started {
		return fmt.Errorf("agent already started")
	}

	ctx := context.Background()

	// Create resource
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceNameKey.String(a.config.ServiceName),
			semconv.DeploymentEnvironmentKey.String(a.config.Environment),
		),
	)
	if err != nil {
		return fmt.Errorf("failed to create resource: %w", err)
	}

	// Determine OTLP endpoint
	otlpEndpoint := a.config.OtlpEndpoint
	if otlpEndpoint == "" {
		otlpEndpoint = fmt.Sprintf("%s/v1/traces", a.config.ControlPlaneURL)
	}

	// Create OTLP exporter
	headers := make(map[string]string)
	if a.config.AuthKey != "" {
		headers["Authorization"] = fmt.Sprintf("Bearer %s", a.config.AuthKey)
	}

	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(otlpEndpoint),
		otlptracehttp.WithHeaders(headers),
	)
	if err != nil {
		return fmt.Errorf("failed to create exporter: %w", err)
	}

	// Create tracer provider
	tp := trace.NewTracerProvider(
		trace.WithBatcher(exporter,
			trace.WithMaxQueueSize(a.config.Telemetry.MaxQueueSize),
			trace.WithBatchTimeout(a.config.Telemetry.BatchTimeout),
			trace.WithExportTimeout(a.config.Telemetry.ExportTimeout),
		),
		trace.WithResource(res),
	)

	otel.SetTracerProvider(tp)
	a.tracerProvider = tp
	a.started = true

	return nil
}

// Stop shuts down the agent.
func (a *Agent) Stop() error {
	if !a.started {
		return nil
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := a.tracerProvider.Shutdown(ctx); err != nil {
		return fmt.Errorf("failed to shutdown tracer provider: %w", err)
	}

	a.started = false
	return nil
}

// GetTracer returns a tracer for manual instrumentation.
func (a *Agent) GetTracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// Redact redacts PII from data.
func (a *Agent) Redact(data interface{}) interface{} {
	return a.redactor.Redact(data)
}

// CheckPolicy checks if an action is allowed by policy.
func (a *Agent) CheckPolicy(action string, context map[string]interface{}) *PolicyResult {
	return a.policyEngine.Check(action, context)
}

// IsStarted returns whether the agent is started.
func (a *Agent) IsStarted() bool {
	return a.started
}
